src/main/java/org/mockbukkit/mockbukkit/entity/VillagerMock.java

```diff
@@ -0,0 +1,251 @@
+		this.level = level;
+	}
+
+	@Override
+	public int getVillagerExperience()
+	{
+		return this.experienceLevel;
+	}
+
+	@Override
+	public void setVillagerExperience(int experience)
+	{
+		Preconditions.checkArgument(experience >= 0, "Experience (%s) must be positive", experience);
+		this.experienceLevel = experience;
+	}
+
+	@Override
+	public boolean increaseLevel(int amount)
+	{
+		Preconditions.checkArgument(amount > 0, "Level earned must be positive");
+		int supposedFinalLevel = this.getVillagerLevel() + amount;
+		Preconditions.checkArgument(MIN_VILLAGER_LEVEL <= supposedFinalLevel && supposedFinalLevel <= MAX_VILLAGER_LEVEL,
+				"Final level reached after the donation (%d) must be between [%d, %d]".formatted(supposedFinalLevel, MIN_VILLAGER_LEVEL, MAX_VILLAGER_LEVEL));
+
+		List<MerchantRecipe> trades = getRecipes();
+		if (trades.isEmpty()) {
+			setVillagerLevel(supposedFinalLevel);
+			return false;
+		}
+
+		while (amount > 0) {
+			setVillagerLevel(getVillagerLevel() + 1);
+			updateTrades();
+			amount--;
+		}
+		return true;
+	}
+
+	@Override
+	public boolean addTrades(int amount)
+	{
+		Preconditions.checkArgument(amount > 0, "Number of trades unlocked must be positive");
+		return updateTrades(amount);
+	}
+
+	@Override
+	public int getRestocksToday()
+	{
+		return this.numberOfRestocksToday;
+	}
+
+	@Override
+	public void setRestocksToday(int restocksToday)
+	{
+		this.numberOfRestocksToday = restocksToday;
+	}
+
+	@Override
+	public boolean sleep(@NotNull Location location)
+	{
+		Preconditions.checkArgument(location != null, "Location cannot be null");
+		Preconditions.checkArgument(location.getWorld() != null, "Location needs to be in a world");
+		Preconditions.checkArgument(location.getWorld().equals(this.getWorld()), "Cannot sleep across worlds");
+
+		if (!MaterialTags.BEDS.isTagged(location.getBlock()))
+		{
+			return false;
+		}
+
+		setSleeping(true);
+
+		leaveVehicle();
+
+		setPose(Pose.SLEEPING);
+		setLocation(location);
+
+		setMemory(MemoryKey.LAST_SLEPT, getWorld().getGameTime());
+		return true;
+	}
+
+	@Override
+	public void wakeup()
+	{
+		Preconditions.checkState(this.isSleeping(), "Cannot wakeup if not sleeping");
+		Preconditions.checkState(getWorld() != null, "Villager needs to be in a world");
+
+		setSleeping(false);
+
+		setPose(Pose.STANDING);
+		// TODO: Set the wakeup location
+
+		setMemory(MemoryKey.LAST_WOKEN, getWorld().getGameTime());
+	}
+
+	@Override
+	public void shakeHead()
+	{
+		// TODO:
+		throw new UnimplementedOperationException();
+	}
+
+	@Override
+	public @Nullable ZombieVillager zombify()
+	{
+		// TODO:
+		throw new UnimplementedOperationException();
+	}
+
+	@Override
+	public @NotNull Reputation getReputation(@NotNull UUID uniqueId)
+	{
+		Reputation value = this.gossips.get(uniqueId);
+		return (value == null ? new Reputation() : value);
```

Comment:
I think the value is copied in the paper implementation. (Reputation is mutable)